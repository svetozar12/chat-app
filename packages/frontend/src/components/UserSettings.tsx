import React from "react";
import { requestUrl } from "../utils/hostUrl_requestUrl";
import { useDispatch } from "react-redux";
import axios from "axios";
import { useRouter } from "next/router";
import Link from "next/link";
import styled from "@emotion/styled";

const User_settings = styled.div`
  width: auto;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  flex-direction: column;
  background: var(--main-white);
  text-align: left;
  left: 0;
  margin: 0 1rem;
  box-shadow: 0 0 10px black;
`;

const User_settings_anchor = styled.a`
  color: black;
  margin: 1rem;
  width: 100%;
  font-size: 1.2vw;
  cursor: pointer;
  &:hover {
    color: rgba(122, 122, 122, 1);
  }
`;
function UserSettings({ cookie }: { cookie: any }) {
  const router = useRouter();
  const dispatch = useDispatch();

  const deleteCookies = () => {
    cookie.remove("name");
    cookie.remove("token");
    cookie.remove("refresh_token");
    router.push("/");
    dispatch({ type: "SIGN_OUT" });
  };

  const deleteUser = async () => {
    try {
      await axios.delete(`${requestUrl}/users/${cookie.get("name")}`);

      deleteCookies();
      return true;
    } catch (error) {
      return false;
    }
  };

  return (
    <User_settings data-testid="user_settings">
      <Link href="/settings/profile">
        <User_settings_anchor
          onClick={() => {
            dispatch({ type: "SET_USER_SETTINGS", payload: false });
          }}
        >
          User settings
        </User_settings_anchor>
      </Link>
      <Link href="#">
        <User_settings_anchor onClick={deleteCookies}>
          Log out
        </User_settings_anchor>
      </Link>
      <Link href="#">
        <User_settings_anchor onClick={deleteUser}>
          Delete user
        </User_settings_anchor>
      </Link>
    </User_settings>
  );
}

export default UserSettings;
