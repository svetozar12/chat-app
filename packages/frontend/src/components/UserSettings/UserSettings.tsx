import React, { useRef } from "react";
import { useDispatch } from "react-redux";
import { useRouter } from "next/router";
import { useCookie } from "next-cookie";
import Link from "next/link";
// utils
import api_helper from "../../services/graphql/api_helper";
import { getAuth } from "../../utils/authMethods";
import { css } from "@emotion/css";
import styled from "@emotion/styled";
import { Link as ALink, VStack } from "@chakra-ui/react";
// icons
import { IoMdLogOut } from "react-icons/io";
import { RiDeleteBin6Fill } from "react-icons/ri";
import { FiSettings } from "react-icons/fi";

const buttonStyles = css`
  margin: 0 1rem;
  border-radius: unset;
  width: 1.5rem;
  height: 1.5rem;
`;

function UserSettings() {
  const router = useRouter();
  const dispatch = useDispatch();
  const cookie = useCookie();
  const deleteCookies = async () => {
    getAuth();
    const cookies = cookie.getAll();
    await api_helper.auth.logout(cookie.get("id"), cookie.get("token"));
    for (const key in cookies) cookie.remove(key);

    router.push("/");
    dispatch({ type: "SIGN_OUT" });
  };

  const deleteUser = async () => {
    try {
      getAuth();
      await api_helper.user.delete(cookie.get("id"), cookie.get("token"));
      deleteCookies();
      return true;
    } catch (error) {
      return false;
    }
  };

  const Render = [
    {
      href: "/settings/profile",
      onClick: () => {
        dispatch({ type: "SET_USER_SETTINGS", payload: false });
      },
      Icon: FiSettings,
      title: "User settings",
    },
    {
      href: "#",
      onClick: deleteCookies,
      Icon: IoMdLogOut,
      title: "Log out",
    },
    {
      href: "#",
      onClick: deleteUser,
      Icon: RiDeleteBin6Fill,
      title: " Delete user",
    },
  ];

  return (
    <VStack
      w={{ base: "15rem", lg: "20rem" }}
      p="1rem"
      pos="absolute"
      justifyContent="center"
      align="center"
      bg="main_white"
      borderRadius="xl"
      textAlign="left"
      top={0}
      right={0}
      m={0}
      zIndex={2000}
      transform="translate(-4px, 75px)"
      boxShadow=" 2px 2px 22px 1px var(--main-box-shadow)"
      _before={{
        pos: "absolute",
        content: `""`,
        height: "20px",
        width: "20px",
        background: "main_white",
        right: " 14px",
        top: "-10px",
        transform: "rotate(45deg)",
        zIndex: "2000",
      }}
    >
      {Render.map((element, index) => {
        const { href, onClick, Icon, title } = element;

        return (
          <Link key={index} href={href} passHref>
            <ALink
              display="flex"
              flexDirection="column"
              alignItems="center"
              textAlign="start"
              w="full"
              borderRadius="xl"
              my="2"
              py="2"
              fontSize={{ base: "1.4rem", md: "3vw", lg: "2vw", xl: "1vw" }}
              _hover={{ bg: "hover_bg" }}
              onClick={onClick}
            >
              <Icon className={buttonStyles} />
              {title}
            </ALink>
          </Link>
        );
      })}
    </VStack>
  );
}

export default UserSettings;
