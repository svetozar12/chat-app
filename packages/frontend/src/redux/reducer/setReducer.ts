import { Action, ActionType, InitialState2 } from "../state";
export const initialState = {
  reciever: "",
  chat_inviter: "",
  pageNumber: 2,
  setUserSettings: false,
  setFriendRequest: false,
  toggleCreateGroup: false,
  setMobileNav: false,
  setIsMatch: false,
  setChatSettings: false,
  setModalInvite: false,
  setIsLoggingIn: false,
};
const setReducer = (state: InitialState2 | any | InitialState2[] = initialState, action: Action) => {
  switch (action.type) {
    case ActionType.SET_RECIEVER:
      return { ...state, reciever: action.payload };
    case ActionType.CHAT_INVITER:
      return { ...state, chat_inviter: action.payload };
    case ActionType.INCREMENT_PAGE_NUMBER:
      return { ...state, pageNumber: ++action.payload };
    case ActionType.SET_FRIEND_REQUEST:
      return { ...state, setFriendRequest: action.payload };
    case ActionType.SET_USER_SETTINGS:
      return { ...state, setUserSettings: action.payload };
    case ActionType.TOGGLE_CREATE_GROUP:
      return { ...state, toggleCreateGroup: action.payload };
    case ActionType.SET_MOBILE_NAV:
      return { ...state, setMobileNav: action.payload };
    case ActionType.SET_CHAT_SETTINGS:
      return { ...state, setChatSettings: action.payload };
    case ActionType.SET_IS_MATCH:
      return { ...state, setIsMatch: action.payload };
    case ActionType.SET_MODAL_INVITE:
      return { ...state, setModalInvite: action.payload };
    case ActionType.SET_IS_LOGEDIN:
      return { ...state, setIsLoggingIn: action.payload };
    case ActionType.SIGN_OUT:
      return initialState;
    default:
      return state;
  }
};

export default setReducer;
