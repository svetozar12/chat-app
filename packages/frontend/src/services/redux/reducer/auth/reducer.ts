import { HYDRATE } from 'next-redux-wrapper';
import { IAuth } from './state';
import { ActionType } from '../../types';

export const initialState: IAuth = {
  remember_me: false,
  loginPrompt: false,
};
const authReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default authReducer;
