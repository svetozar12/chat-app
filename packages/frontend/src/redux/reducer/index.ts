import { combineReducers } from "redux";
import reducer from "./authReducer";
import setReducer from "./setReducer";
import saveInputReducer from "./save_inputReducer";
import messageReducer from "./messageReducer";

const reducers = combineReducers({
  authReducer: reducer,
  setReducer,
  messageReducer,
  saveInputReducer,
});

export default reducers;
