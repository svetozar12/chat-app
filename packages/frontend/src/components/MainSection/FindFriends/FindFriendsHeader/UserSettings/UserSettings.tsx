import React from 'react';
import { connect } from 'react-redux';
import { useRouter } from 'next/router';
import { useCookie } from 'next-cookie';
import Link from 'next/link';
// utils
import { css } from '@emotion/css';
import { Link as ALink, VStack } from '@chakra-ui/react';
// icons
import { IoMdLogOut } from 'react-icons/io';
import { RiDeleteBin6Fill } from 'react-icons/ri';
import { FiSettings } from 'react-icons/fi';
import { getAuth } from '../../../../../utils/authMethods';
import useThemeColors from '../../../../../hooks/useThemeColors';
import { bindActionCreators, Dispatch } from 'redux';
import { toggleUserSettings } from 'services/redux/reducer/toggles/actions';
import { signOut } from 'services/redux/reducer/auth/actions';
import sdk from 'services/sdk';

const buttonStyles = css`
  margin: 0 1rem;
  border-radius: unset;
  width: 1.5rem;
  height: 1.5rem;
`;

interface IUserSettings {
  toggleUserSettings: typeof toggleUserSettings;
  signOut: typeof signOut;
}

function UserSettings(props: IUserSettings) {
  const { toggleUserSettings, signOut } = props;
  const router = useRouter();
  const cookie = useCookie();

  const deleteCookies = async () => {
    getAuth();
    const cookies = cookie.getAll();
    await sdk.auth.logout({ auth: { userId: cookie.get('id'), AccessToken: cookie.get('token') } });
    for (const key in cookies) cookie.remove(key);

    router.push('/');
    signOut();
  };

  const deleteUser = async () => {
    try {
      getAuth();
      await sdk.user.deleteUser({ auth: { userId: cookie.get('id'), AccessToken: cookie.get('token') } });
      deleteCookies();
      return true;
    } catch (error) {
      return false;
    }
  };

  const Render = [
    {
      href: '/settings/profile',
      onClick: () => {
        toggleUserSettings(false);
      },
      Icon: FiSettings,
      title: 'User settings',
    },
    {
      href: '#',
      onClick: deleteCookies,
      Icon: IoMdLogOut,
      title: 'Log out',
    },
    {
      href: '#',
      onClick: deleteUser,
      Icon: RiDeleteBin6Fill,
      title: ' Delete user',
    },
  ];

  const {
    colors: { fromBg },
  } = useThemeColors();
  return (
    <VStack
      w={{ base: '15rem', lg: '15rem' }}
      pos="absolute"
      justifyContent="center"
      align="center"
      bg={fromBg}
      textAlign="left"
      top={0}
      right={0}
      m={0}
      pt={2}
      pb={2}
      zIndex={2000}
      transform="translate(-4px, 75px)"
      boxShadow=" 2px 2px 22px 1px var(--main-box-shadow)"
    >
      {Render.map((element, index) => {
        const { href, onClick, Icon, title } = element;

        return (
          <Link key={index} href={href} passHref>
            <ALink
              display="flex"
              justifyContent="flex-start"
              alignItems="center"
              textAlign="start"
              w="full"
              my="0 !important"
              py="2"
              fontSize={{ base: '1.4rem', md: '3vw', lg: '2vw', xl: '1vw' }}
              _hover={{ bg: 'hover_bg' }}
              onClick={onClick}
            >
              <Icon className={buttonStyles} />
              {title}
            </ALink>
          </Link>
        );
      })}
    </VStack>
  );
}

const mapDispatchToProps = (dispatch: Dispatch) => ({
  toggleUserSettings: bindActionCreators(toggleUserSettings, dispatch),
  signOut: bindActionCreators(signOut, dispatch),
});

export default connect(null, mapDispatchToProps)(UserSettings);
