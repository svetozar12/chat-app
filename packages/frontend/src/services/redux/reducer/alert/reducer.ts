import { AnyAction } from 'redux';
import { IAlert } from 'services/redux/reducer/alert/state';
import { ActionType } from '../../types';

const initialState: IAlert = {
  good: '', // good and bad stand for good alert and bad alert
  bad: '',
};

const alertReducer = (state = initialState, action: AnyAction) => {
  switch (action.type) {
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.payload };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.payload };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.payload };
    default:
      return state;
  }
};

export default alertReducer;
