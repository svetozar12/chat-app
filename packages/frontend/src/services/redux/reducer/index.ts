import { AnyAction, combineReducers } from 'redux';
import authReducer from 'services/redux/reducer/auth/reducer';
import { ActionType } from '../types';
import { HYDRATE } from 'next-redux-wrapper';
import wsReducer from 'services/redux/reducer/websocket/reducer';
import messageReducer from 'services/redux/reducer/messages/reducer';

import { IAuth } from 'services/redux/reducer/auth/state';
import { IWebSocket } from 'services/redux/reducer/websocket/state';
import { IMessage } from 'services/redux/reducer/messages/state';
import IInputs from 'services/redux/reducer/inputs/state';
import inputReducer from 'services/redux/reducer/inputs/reducer';
import toggleReducer from 'services/redux/reducer/toggles/reducer';
import { IToggle } from 'services/redux/reducer/toggles/state';
import IInvite from 'services/redux/reducer/invites/state';
import inviteReducer from 'services/redux/reducer/invites/reducer';
import { IAlert } from 'services/redux/reducer/alert/state';
import alertReducer from 'services/redux/reducer/alert/reducer';

export interface STATE {
  auth: IAuth;
  ws: IWebSocket;
  messages: IMessage;
  inputs: IInputs;
  toggle: IToggle;
  invite: IInvite;
  alert: IAlert;
}

const combiReducers = combineReducers({
  auth: authReducer,
  ws: wsReducer,
  messages: messageReducer,
  inputs: inputReducer,
  toggle: toggleReducer,
  invite: inviteReducer,
  alert: alertReducer,
});

const combReducers = (state: any, action: AnyAction) => {
  if (action.type === HYDRATE)
    return {
      ...state,
      ...action.payload,
    };
  if (action.type === ActionType.SIGN_OUT) {
    return undefined;
  }
  return combiReducers(state, action);
};

export default combReducers;
