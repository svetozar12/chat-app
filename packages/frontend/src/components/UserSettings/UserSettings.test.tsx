import "@testing-library/jest-dom";
import { cleanup, render, screen } from "@testing-library/react";
import UserSettings from "./UserSettings";
import { Provider } from "react-redux";
import { initialState } from "redux/reducer/setReducer";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
let container;
beforeAll(() => {
  const mockStore = configureStore([]);
  const store = mockStore({
    setReducer: initialState,
  });
  container = renderer
    .create(
      <Provider store={store}>
        <UserSettings cookie="test2" />
      </Provider>,
    )
    .toJSON();
  render(
    <Provider store={store}>
      <UserSettings cookie="test2" />
    </Provider>,
  );
});

afterAll(cleanup);

describe("<UserSettings/>", () => {
  it("should create snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should render", () => {
    const test = screen.getByTestId("user_settings");
    expect(test).toBeInTheDocument();
  });
});
