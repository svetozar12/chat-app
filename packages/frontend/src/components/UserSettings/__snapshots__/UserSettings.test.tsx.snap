// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSettings/> should create snapshot 1`] = `
<div
  className="css-v8ndqr"
  data-testid="user_settings"
>
  <a
    className="css-jh977x"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    User settings
  </a>
  <a
    className="css-jh977x"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    Log out
  </a>
  <a
    className="css-jh977x"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    Delete user
  </a>
</div>
`;
